locator/ec2_snitch.cc

```diff
@@ -85,6 +95,12 @@ future<sstring> ec2_snitch::aws_api_call(sstring addr, uint16_t port, sstring cm
 
             // Read HTTP response header first
             auto _rsp = _parser.get_parsed_response();
+            auto rc = _rsp->_status_code;
+            // Verify EC2 instance metadata access
+            if (rc == 403) {
+                throw std::runtime_error("Error: Unauthorized response received when trying to communicate with instance metadata service.");
```

Comment:
Better use `return make_exception_future<std::runtime_error>(...)` instead, like we do in line 106 below.

The final result will be the same but it makes sense to keep the code homogeneous.

Reply:
I assume you mean `make_exception_future<sstring>(std::runtime_error(...))`. Simply returning a `sstring` exception causes the startup to fail with:

```shell
init - Startup failed: char const*
```

Which is worse than directly throwing - as directly throwing properly tells us what is wrong.

My assumption is that we've simply never tested these exceptions. So I can change it to `make_exception_future<sstring>(std::runtime_error(...))` as mentioned, and it will have the same effect, however it won't keep the code homogeneous IMO, so let me know if you want me to also correct the remaining exceptions. :-)

Reply:
> I assume you mean make_exception_future<sstring>(std::runtime_error(...))

Correct, that's what I was supposed to "mean" - I forgot the semantics of the `make_exception_future` when I wrote the previous comment.

The template type is supposed to match the caller return future type (`sstring` in this case) and the argument can be anything that can be `throw`n in C++, which is technically any object.

So, `make_exception_future<sstring>(std::runtime_error(...))` should work nicely.
So, should `make_exception_future<sstring>(sstring(...))`. 
More than that, throwing "const char*" is ok too:

```
test.cc:

#include <iostream>

int main() {
        try {
                throw "Throwing char *";
        } catch (const char * ch) {
                std::cout << ch;
        }
        return 0;
}


$ ./test
Throwing char *
```

> So I can change it to make_exception_future<sstring>(std::runtime_error(...)) as mentioned, and it will have the same effect, however it won't keep the code homogeneous IMO

I think it would. If you imply that wrapping the text with a `std::runtime_error` makes it "non-homogeneous" - then I'd argue that this way you tag the corresponding error with a type ("runtime error" in this case) which is absolutely fine and is not in conflict with the existing code which doesn't add such a tag.

Reply:
I suspect that the issue with the printing is related to the challenges related to printing `std::exception_ptr` in the current code (we rely on `fmt` now) which is done by throwing and then catching "known" types with apparently includes `std::exception` but doesn't include `const char*`. 

@avikivity Could you, please, confirm?

Reply:
Right, I have already pushed an update since yesterday with a call to `make_exception_future` rather than directly throwing, so unless you want me to update the other seemingly broken `make_exception_future<sstring>` calls not relevant to the main purpose of this commit, let me know, otherwise I think we can flag this conversation as resolved. :-)

Reply:
While C++ technically allows you to throw any type (which is a mistake IMO) if you want your exception to be properly handled it should be a type that inherits from `std::exception`. So please don't throw `sstring` directly, it might look like a good idea but it is really not.

Reply:
> While C++ technically allows you to throw any type (which is a mistake IMO) if you want your exception to be properly handled it should be a type that inherits from std::exception. So please don't throw sstring directly, it might look like a good idea but it is really not.

Since we know that the current code base definitely doesn't support throwing `char *` (I have a reason to believe it was supported before because I usually test my patches ;)) it probably makes sense to send a follow up that fixes other `make_exception_future` in this file.

And like Felipe already mentioned this clearly exposes a hole in our QA testing in regards to "bad case" testing.